backend/core/src/main/java/org/sonarsource/sonarlint/core/ConnectionManager.java

```diff
@@ -145,24 +148,26 @@ private HttpClient getClientFor(EndpointParams params, Either<TokenDto, Username
    * Throws ResponseErrorException if connection with provided ID is not found in ConnectionConfigurationRepository
    */
   public ServerConnection getConnectionOrThrow(String connectionId) {
-    var serverApi = getServerApiOrThrow(connectionId);
-    return new ServerConnection(connectionId, serverApi, client);
+    return connectionCache.computeIfAbsent(connectionId, connId -> {
+      var serverApi = getServerApiOrThrow(connId);
+      return new ServerConnection(connId, serverApi, client);
+    });
   }
 
   /**
    * Returns empty Optional if connection with provided ID is not found in ConnectionConfigurationRepository
    */
   public Optional<ServerConnection> tryGetConnection(String connectionId) {
-    return getServerApi(connectionId)
-      .map(serverApi -> new ServerConnection(connectionId, serverApi, client));
+    return Optional.ofNullable(connectionCache.computeIfAbsent(connectionId, connId ->
+      getServerApi(connId).map(serverApi -> new ServerConnection(connId, serverApi, client)).orElse(null)));
   }
 
   /**
    * Should be used for WebAPI requests without an authentication
    */
   public Optional<ServerConnection> tryGetConnectionWithoutCredentials(String connectionId) {
     return getServerApiWithoutCredentials(connectionId)
-      .map(serverApi -> new ServerConnection(connectionId, serverApi, client));
+      .map(serverApi -> new ServerConnection(connectionId, serverApi, client, true));
```

Comment:
Similarly, I think this method should be removed. The only use case is when we want to fetch the server version to check if it is soon unsupported. I don't see why we wouldn't make an authenticated call. If this specific endpoint does not require authentication, then it's more at the `SystemApi` level that we should not pass the credentials when calling `get`, maybe we need a `getAnonymous` in the ServerApiHelper. This would simplify the new constructor you added

Reply:
Same. See other comment.